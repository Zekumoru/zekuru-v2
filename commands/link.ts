import {
  ChatInputCommandInteraction,
  PermissionFlagsBits,
  SlashCommandBuilder,
} from 'discord.js';
import { createCommand } from '../types/DiscordCommand';
import translateChannels from '../cache/translateChannels';
import channelLinks from '../cache/channelLinks';
import { IChannelLink } from '../models/ChannelLink';
import { ITranslateChannel } from '../models/TranslateChannel';

export const CHANNEL_LINK_LIMIT = isNaN(Number(process.env.CHANNEL_LINK_LIMIT))
  ? 5
  : Number(process.env.CHANNEL_LINK_LIMIT);

const LinkOptions = {
  SOURCE_CHANNEL: 'source-channel',
  TARGET_CHANNEL: 'target-channel',
  MODE: 'mode',
  mode: {
    UNIDIRECTIONAL: 'unidirectional',
    BIDIRECTIONAL: 'bidirectional',
    RECURSIVE: 'recursive',
  },
};

const data = new SlashCommandBuilder()
  .setName('link')
  .setDescription('Links two translation channels.')
  .addChannelOption((option) =>
    option
      .setName(LinkOptions.TARGET_CHANNEL)
      .setDescription('The target channel to link.')
      .setRequired(true)
  )
  .addChannelOption((option) =>
    option
      .setName(LinkOptions.SOURCE_CHANNEL)
      .setDescription(
        'The source channel to link. If not provided, takes the current channel as the source.'
      )
  )
  .addStringOption((option) =>
    option
      .setName(LinkOptions.MODE)
      .setDescription('Specify linking mode. Default is bidirectional.')
      .addChoices(
        {
          name: LinkOptions.mode.UNIDIRECTIONAL,
          value: LinkOptions.mode.UNIDIRECTIONAL,
        },
        {
          name: LinkOptions.mode.BIDIRECTIONAL,
          value: LinkOptions.mode.BIDIRECTIONAL,
        },
        {
          name: LinkOptions.mode.RECURSIVE,
          value: LinkOptions.mode.RECURSIVE,
        }
      )
  )
  .setDefaultMemberPermissions(PermissionFlagsBits.ManageGuild);

export const getChLink = async (channelId: string, guildId: string) => {
  return (
    (await channelLinks.get(channelId)) ??
    (await channelLinks.create(channelId, guildId))
  );
};

export const linkChannel = async (
  channelLink: IChannelLink,
  channelId: string,
  translateChannel: ITranslateChannel
) => {
  if (channelLink.links.find((link) => link.id === channelId) === undefined) {
    channelLink.links.push(translateChannel);
    await channelLinks.update(channelLink);
    return true;
  }
  return false;
};

export interface IAllChLinkMapValue {
  chLink: IChannelLink;
  trChannel: ITranslateChannel;
}

export const linkChannelNoDB = (
  channelLink: IChannelLink,
  channelId: string,
  translateChannel: ITranslateChannel
) => {
  if (channelLink.links.find((link) => link.id === channelId) === undefined) {
    channelLink.links.push(translateChannel);
    return true;
  }
  return false;
};

export const linkChannels = async (
  allChLinkMap: Map<string, IAllChLinkMapValue>
) => {
  let linked = false;
  allChLinkMap.forEach(({ chLink: sourceChLink }) => {
    allChLinkMap.forEach(
      ({ chLink: targetChLink, trChannel: targetTrChannel }) => {
        if (sourceChLink === targetChLink) return;

        linked =
          linkChannelNoDB(sourceChLink, targetChLink.id, targetTrChannel) ||
          linked;
      }
    );
  });

  // save to db
  const promises: Promise<void>[] = [];
  allChLinkMap.forEach(({ chLink }) => {
    promises.push(
      (async () => {
        await channelLinks.update(chLink);
      })()
    );
  });
  await Promise.all(promises);

  return linked;
};

export const buildAllChannelsLinkMap = async (channelLinks: IChannelLink[]) => {
  const jobQueue: IChannelLink[] = [...channelLinks];
  const allChLinkMap = new Map<string, IAllChLinkMapValue>();

  // build map containing all channels
  while (jobQueue.length) {
    const chLink = jobQueue.shift()!;

    for (const link of chLink.links) {
      if (allChLinkMap.get(link.id)) continue;

      const [linkTrChannel, linkChLink] = await Promise.all([
        translateChannels.get(link.id),
        getChLink(link.id, chLink.guildId),
      ]);
      jobQueue.push(linkChLink);

      if (!linkTrChannel) continue;

      allChLinkMap.set(linkChLink.id, {
        chLink: linkChLink,
        trChannel: linkTrChannel,
      });
    }
  }

  return allChLinkMap;
};

const execute = async (interaction: ChatInputCommandInteraction) => {
  if (!interaction.guildId) {
    await interaction.reply({
      content: `This command is only available on servers.`,
    });
    return;
  }

  const sourceChannelId =
    interaction.options.getChannel(LinkOptions.SOURCE_CHANNEL)?.id ??
    interaction.channelId;

  const targetChannelId = interaction.options.getChannel(
    LinkOptions.TARGET_CHANNEL
  )?.id;
  if (!targetChannelId) {
    await interaction.reply({
      content: `Please specify the target channel.`,
    });
    return;
  }

  const mode =
    interaction.options.getString(LinkOptions.MODE) ??
    LinkOptions.mode.BIDIRECTIONAL;

  // check if these channels have languages associated with them
  const [sourceTrChannel, targetTrChannel] = await Promise.all([
    translateChannels.get(sourceChannelId),
    translateChannels.get(targetChannelId),
  ]);

  let errorMessage = '';
  if (sourceChannelId === targetChannelId) {
    errorMessage = `You cannot link <#${sourceChannelId}> with itself!`;
  } else if (sourceTrChannel == null && targetTrChannel == null) {
    errorMessage = `Both <#${sourceChannelId}> and <#${targetChannelId}> are not associated with any languages yet. Please use the \`/set\` command to set their languages.`;
  } else if (sourceTrChannel == null) {
    errorMessage = `<#${sourceChannelId}> is not associated with any languages yet. Please use the \`/set\` command to set its language.`;
  } else if (targetTrChannel == null) {
    errorMessage = `<#${targetChannelId}> is not associated with any languages yet. Please use the \`/set\` command to set its language.`;
  }

  if (errorMessage || sourceTrChannel == null || targetTrChannel == null) {
    await interaction.reply({
      content: errorMessage,
    });
    return;
  }

  // add to their respective link documents
  const [sourceChLink, targetChLink] = await Promise.all([
    getChLink(sourceChannelId, interaction.guildId),
    getChLink(targetChannelId, interaction.guildId),
  ]);

  const allChLinkMap = await buildAllChannelsLinkMap([
    sourceChLink,
    targetChLink,
  ]);

  // don't link if reached linking limit
  if (allChLinkMap.size >= CHANNEL_LINK_LIMIT) {
    await interaction.reply({
      content: `<#${sourceChannelId}> **(${sourceTrChannel.sourceLang})** and <#${targetChannelId}> **(${targetTrChannel.sourceLang})** are **not linked** because you already reached the linking limit of ${CHANNEL_LINK_LIMIT}!`,
    });
    return;
  }

  let linked = false;

  // link unidirectional
  linked = await linkChannel(sourceChLink, targetChannelId, targetTrChannel);
  if (mode === LinkOptions.mode.UNIDIRECTIONAL) {
    await interaction.reply({
      content: linked
        ? `<#${sourceChannelId}> **(${sourceTrChannel.sourceLang})** is now linked **unidirectionally** to <#${targetChannelId}> **(${targetTrChannel.sourceLang})**!`
        : `<#${sourceChannelId}> **(${sourceTrChannel.sourceLang})** is already linked with <#${targetChannelId}> **(${targetTrChannel.sourceLang})**!`,
    });
    return;
  }

  // link bidirectional
  linked =
    (await linkChannel(targetChLink, sourceChannelId, sourceTrChannel)) ||
    linked;
  if (mode === LinkOptions.mode.BIDIRECTIONAL) {
    await interaction.reply({
      content: linked
        ? `<#${sourceChannelId}> **(${sourceTrChannel.sourceLang})** and <#${targetChannelId}> **(${targetTrChannel.sourceLang})** are now linked!`
        : `<#${sourceChannelId}> **(${sourceTrChannel.sourceLang})** and <#${targetChannelId}> **(${targetTrChannel.sourceLang})** are already linked!`,
    });
    return;
  }

  // link recursively
  // already add these to map to save time since they're already done as well
  // refetch channels links due to bidirectional linking
  allChLinkMap.set(sourceChLink.id, {
    chLink: await getChLink(sourceChannelId, interaction.guildId),
    trChannel: sourceTrChannel,
  });
  allChLinkMap.set(targetChLink.id, {
    chLink: await getChLink(targetChannelId, interaction.guildId),
    trChannel: targetTrChannel,
  });

  // start linking recursively O(n^2) [actually O(n^3) because of linkChannel()]
  linked = (await linkChannels(allChLinkMap)) || linked;

  await interaction.reply({
    content: linked
      ? `<#${sourceChannelId}> **(${sourceTrChannel.sourceLang})** and <#${targetChannelId}> **(${targetTrChannel.sourceLang})** are now linked **recursively**!`
      : `<#${sourceChannelId}> **(${sourceTrChannel.sourceLang})** and <#${targetChannelId}> **(${targetTrChannel.sourceLang})** are already linked **recursively**!`,
  });
};

export default createCommand({
  data,
  execute,
});
