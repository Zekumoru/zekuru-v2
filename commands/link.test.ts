import link, { LinkOptions } from './link';
import {
  clearSampleChannelLinks,
  sampleChannelLinks,
  sampleTranslateChannels,
} from '../cache/__mocks__/sample-data';
import createMockInteraction from '../test/createMockChatInputCommandInteraction';

jest.mock('./utilities/linking');
jest.mock('../cache');

describe('link', () => {
  afterEach(() => {
    clearSampleChannelLinks();
    jest.clearAllMocks();
  });

  it('should notify user that the command has been called outside of a server', async () => {
    const interaction = createMockInteraction({ guildId: undefined });

    await link.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if they didn't provide a translate channel`, async () => {
    const interaction = createMockInteraction({ channelId: '100' });

    await link.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if they tried to link a channel to itself`, async () => {
    const interaction = createMockInteraction({ channelId: '100' });
    interaction.options.getChannel.mockReturnValueOnce(null);
    interaction.options.getChannel.mockReturnValueOnce({ id: '100' } as any);

    await link.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if they provided channels that aren't translate channels`, async () => {
    const interaction = createMockInteraction({ channelId: '000' });
    interaction.options.getChannel.mockReturnValueOnce(null);
    interaction.options.getChannel.mockReturnValueOnce({ id: '001' } as any);

    await link.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if they provided a source channel that is not a translate channel`, async () => {
    const interaction = createMockInteraction({ channelId: '000' });
    interaction.options.getChannel.mockReturnValueOnce(null);
    interaction.options.getChannel.mockReturnValueOnce({ id: '100' } as any);

    await link.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should notify user if they provided a target channel that is not a translate channel`, async () => {
    const interaction = createMockInteraction({ channelId: '100' });
    interaction.options.getChannel.mockReturnValueOnce(null);
    interaction.options.getChannel.mockReturnValueOnce({ id: '000' } as any);

    await link.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  it(`should link the channel where the command is called on and the provided target channel`, async () => {
    const interaction = createMockInteraction({ channelId: '100' });
    interaction.options.getChannel.mockReturnValueOnce(null);
    interaction.options.getChannel.mockReturnValueOnce({ id: '200' } as any);

    await link.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    expect(sampleChannelLinks[0].links).toEqual([sampleTranslateChannels[1]]);
    expect(sampleChannelLinks[1].links).toEqual([sampleTranslateChannels[0]]);
  });

  it(`should link the provided source channel and the provided target channel`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getChannel.mockReturnValueOnce({ id: '100' } as any);
    interaction.options.getChannel.mockReturnValueOnce({ id: '200' } as any);

    await link.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    expect(sampleChannelLinks[0].links).toEqual([sampleTranslateChannels[1]]);
    expect(sampleChannelLinks[1].links).toEqual([sampleTranslateChannels[0]]);
  });

  it(`should not link if over the linking limit`, async () => {
    sampleChannelLinks[0].links = [
      sampleTranslateChannels[1],
      sampleTranslateChannels[2],
    ] as any;
    sampleChannelLinks[1].links = [
      sampleTranslateChannels[0],
      sampleTranslateChannels[2],
    ] as any;
    sampleChannelLinks[2].links = [
      sampleTranslateChannels[0],
      sampleTranslateChannels[1],
    ] as any;
    const interaction = createMockInteraction();
    interaction.options.getChannel.mockReturnValueOnce({ id: '100' } as any);
    interaction.options.getChannel.mockReturnValueOnce({ id: '400' } as any);

    await link.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
  });

  test(`that unidirectional linking works`, async () => {
    const interaction = createMockInteraction();
    interaction.options.getChannel.mockReturnValueOnce({ id: '100' } as any);
    interaction.options.getChannel.mockReturnValueOnce({ id: '200' } as any);
    interaction.options.getString.mockReturnValue(
      LinkOptions.mode.UNIDIRECTIONAL
    );

    await link.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    expect(sampleChannelLinks[0].links).toContain(sampleTranslateChannels[1]);
    expect(sampleChannelLinks[1].links).not.toContain(
      sampleTranslateChannels[0]
    );
  });

  it(`should link if near the linking limit with recursive linking mode`, async () => {
    sampleChannelLinks[0].links = [sampleTranslateChannels[1]] as any;
    sampleChannelLinks[1].links = [sampleTranslateChannels[0]] as any;
    const interaction = createMockInteraction();
    interaction.options.getChannel.mockReturnValueOnce({ id: '100' } as any);
    interaction.options.getChannel.mockReturnValueOnce({ id: '300' } as any);
    interaction.options.getString.mockReturnValue(LinkOptions.mode.RECURSIVE);

    await link.execute(interaction);

    expect(interaction.reply.mock.calls[0][0]).toMatchSnapshot();
    expect(sampleChannelLinks[0].links).toContain(sampleTranslateChannels[2]);
    expect(sampleChannelLinks[1].links).toContain(sampleTranslateChannels[2]);
    expect(sampleChannelLinks[2].links).toEqual([
      sampleTranslateChannels[0],
      sampleTranslateChannels[1],
    ]);
  });
});
