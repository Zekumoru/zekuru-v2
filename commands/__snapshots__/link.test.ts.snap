// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`link should link if near the linking limit with recursive linking mode 1`] = `
{
  "content": "<#100> **(en)** and <#300> **(zh)** are now linked **recursively**!",
}
`;

exports[`link should link the channel where the command is called on and the provided target channel 1`] = `
{
  "content": "<#100> **(en)** and <#200> **(ja)** are now linked!",
}
`;

exports[`link should link the provided source channel and the provided target channel 1`] = `
{
  "content": "<#100> **(en)** and <#200> **(ja)** are now linked!",
}
`;

exports[`link should not link if over the linking limit 1`] = `
{
  "content": "<#100> **(en)** and <#400> **(ko)** are **not linked** because you already reached the linking limit of 3!",
}
`;

exports[`link should notify user if they didn't provide a translate channel 1`] = `
{
  "content": "Please specify the target channel.",
}
`;

exports[`link should notify user if they provided a source channel that is not a translate channel 1`] = `
{
  "content": "<#000> is not associated with any languages yet. Please use the \`/set\` command to set its language.",
}
`;

exports[`link should notify user if they provided a target channel that is not a translate channel 1`] = `
{
  "content": "<#000> is not associated with any languages yet. Please use the \`/set\` command to set its language.",
}
`;

exports[`link should notify user if they provided channels that aren't translate channels 1`] = `
{
  "content": "Both <#000> and <#001> are not associated with any languages yet. Please use the \`/set\` command to set their languages.",
}
`;

exports[`link should notify user if they tried to link a channel to itself 1`] = `
{
  "content": "You cannot link <#100> with itself!",
}
`;

exports[`link should notify user that the command has been called outside of a server 1`] = `
{
  "content": "This command is only available on servers.",
}
`;

exports[`link that unidirectional linking works 1`] = `
{
  "content": "<#100> **(en)** is now linked **unidirectionally** to <#200> **(ja)**!",
}
`;
